### PR TITLE
[Merged by Bors] - feat: reduction of DirichletCharacter

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 1]
```diff
@@ -182,6 +182,16 @@ lemma conductor_eq_zero_iff_level_eq_zero : conductor χ = 0 ↔ n = 0 := by
   rintro rfl
   exact Nat.sInf_eq_zero.mpr <| Or.inl <| level_mem_conductorSet χ
 
+lemma mem_conductorSet_eq_conductor {d : ℕ} (hd : d ∈ conductorSet χ) :
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:185]: The name of this lemma does not match the statement, Maybe `conductor_le_conductor_mem_conductorSet`?

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 2]
```diff
@@ -199,4 +209,21 @@ lemma conductor_one_dvd (n : ℕ) : conductor (1 : DirichletCharacter R 1) ∣ n
   rw [(isPrimitive_def _).mp isPrimitive_one_level_one]
   apply one_dvd _
 
+/-- The primitive character associated to a Dirichlet character. -/
+noncomputable def reduction : DirichletCharacter R χ.conductor :=
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:213]: I think "reduction" is too generic. Why not `primitiveCharacter` instead?

**Comment by MichaelStollBayreuth** [comment]
Can you replace the `R : Type` by `R : Type*`? (Lines 48, 51, 67.) This would make the target ring universe-polymorphic (and doesn't seem to break anything).

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 2 above)
Comment (laughinggas) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean:213]: `primitiveCharacter` is quite long, I am happy with `PrimitiveChar` or simply `Primitive`

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 1 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: yes, done

**Comment by mo271** [comment]
> Can you replace the `R : Type` by `R : Type*`? (Lines 48, 51, 67.) This would make the target ring universe-polymorphic (and doesn't seem to break anything).

Yeah, good idea, done.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 3]
```diff
@@ -182,6 +182,16 @@ lemma conductor_eq_zero_iff_level_eq_zero : conductor χ = 0 ↔ n = 0 := by
   rintro rfl
   exact Nat.sInf_eq_zero.mpr <| Or.inl <| level_mem_conductorSet χ
 
+lemma conductor_le_conductor_mem_conductorSet {d : ℕ} (hd : d ∈ conductorSet χ) :
+    χ.conductor ≤ (Classical.choose hd.2 ).conductor := by
+  apply Nat.sInf_le
+  rw [mem_conductorSet_iff]
+  refine' ⟨dvd_trans (conductor_dvd_level _) hd.1,
+      (factorsThrough_conductor (Classical.choose hd.2)).2.choose, _⟩
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:190]:
```suggestion
  refine Nat.sInf_le <| (mem_conductorSet_iff χ).mpr <|
    ⟨dvd_trans (conductor_dvd_level _) hd.1,
     (factorsThrough_conductor (Classical.choose hd.2)).2.choose, ?_⟩
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 4]
```diff
@@ -182,6 +182,16 @@ lemma conductor_eq_zero_iff_level_eq_zero : conductor χ = 0 ↔ n = 0 := by
...
+  rw [changeLevel_trans _ (conductor_dvd_level _) (FactorsThrough.dvd _ hd),
+      ← (factorsThrough_conductor (Classical.choose hd.2)).2.choose_spec]
+  apply FactorsThrough.eq_changeLevel χ hd
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:193]:
```suggestion
  exact FactorsThrough.eq_changeLevel χ hd
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 5]
```diff
@@ -199,4 +209,21 @@ lemma conductor_one_dvd (n : ℕ) : conductor (1 : DirichletCharacter R 1) ∣ n
...
+noncomputable def primitiveCharacter : DirichletCharacter R χ.conductor :=
+  Classical.choose (factorsThrough_conductor χ).choose_spec
+
+lemma reduction_isPrimitive : isPrimitive (χ.primitiveCharacter) := by
+  by_cases χ.conductor = 0
+  · rw [isPrimitive_def]
+    conv_rhs => rw [h]
+    rw [conductor_eq_zero_iff_level_eq_zero, h]
+  · refine' le_antisymm (Nat.le_of_dvd (Nat.pos_of_ne_zero h) (conductor_dvd_level _)) <|
+        conductor_le_conductor_mem_conductorSet <| conductor_mem_conductorSet χ
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:222]:
```suggestion
  by_cases h : χ.conductor = 0
  · rw [isPrimitive_def]
    convert conductor_eq_zero_iff_level_eq_zero.mpr h
  · exact le_antisymm (Nat.le_of_dvd (Nat.pos_of_ne_zero h) (conductor_dvd_level _)) <|
      conductor_le_conductor_mem_conductorSet _ <| conductor_mem_conductorSet χ
```
(don't rely on the name `h` used by `by_cases` by default)

Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: I'm happy to change that, but why not rely on the default name?

Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: (The `_` before the last `<|` should go; in my test file I mistakenly had an explicit argument there.)
Relying on the deault name has the problem that a reader of the proof has to know about it (and the possible problem that the proof could break at some point if the default is changed).
In fact, I think that `by_cases` should use an inaccessible name by default, similarly to most other tactics.

**Review by MichaelStollBayreuth — APPROVED** [review]
Looks good to me now.

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 6]
```diff
@@ -199,4 +208,20 @@ lemma conductor_one_dvd (n : ℕ) : conductor (1 : DirichletCharacter R 1) ∣ n
...
+/-- The primitive character associated to a Dirichlet character. -/
+noncomputable def primitiveCharacter : DirichletCharacter R χ.conductor :=
+  Classical.choose (factorsThrough_conductor χ).choose_spec
+
+lemma reduction_isPrimitive : isPrimitive (χ.primitiveCharacter) := by
+  by_cases h : χ.conductor = 0
+  · rw [isPrimitive_def]
+    convert conductor_eq_zero_iff_level_eq_zero.mpr h
+  · exact le_antisymm (Nat.le_of_dvd (Nat.pos_of_ne_zero h) (conductor_dvd_level _)) <|
+      conductor_le_conductor_mem_conductorSet _ <| conductor_mem_conductorSet χ
```
Comment (mo271) [inline Mathlib/NumberTheory/DirichletCharacter/Basic.lean:220]:
```suggestion
      conductor_le_conductor_mem_conductorSet <| conductor_mem_conductorSet χ
```

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean (hunk 5 above)
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: What you write about `by_cases` seems reasonable, I started a discussing about it here:

https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60by_cases.60using.20the.20default.20argument

**Review by laughinggas — COMMENTED** [review]
Seems fine to me!

**Review by laughinggas — APPROVED** [review]

**Review by Vierkantor — APPROVED** [review]
Looks good to me, thanks!

bors d+

File: Mathlib/NumberTheory/DirichletCharacter/Basic.lean [hunk 7]
```diff
@@ -199,4 +208,20 @@ lemma conductor_one_dvd (n : ℕ) : conductor (1 : DirichletCharacter R 1) ∣ n
...
+  by_cases h : χ.conductor = 0
+  · rw [isPrimitive_def]
+    convert conductor_eq_zero_iff_level_eq_zero.mpr h
+  · exact le_antisymm (Nat.le_of_dvd (Nat.pos_of_ne_zero h) (conductor_dvd_level _)) <|
+      conductor_le_conductor_mem_conductorSet <| conductor_mem_conductorSet χ
+
+lemma reduction_one (hn : n ≠ 0) : (1 : DirichletCharacter R n).primitiveCharacter = 1 := by
+  rw [eq_one_iff_conductor_eq_one <| (@conductor_one R _ _ hn) ▸ Nat.one_ne_zero,
+      (isPrimitive_def _).1 (1 : DirichletCharacter R n).reduction_isPrimitive,
+      conductor_one hn]
```
[the remark this answers is not in the document]
Comment (mo271) [reply Mathlib/NumberTheory/DirichletCharacter/Basic.lean]: of course!

**Pull request closed without merging**